cranelift/control/src/chaos.rs

```diff
@@ -0,0 +1,28 @@
+use arbitrary::Arbitrary;
+
+/// The control plane of chaos mode.
+/// Please see the [crate-level documentation](crate).
+#[derive(Debug, Clone, Default)]
+pub struct ControlPlane {
+    data: Vec<bool>,
+}
+
+impl Arbitrary<'_> for ControlPlane {
+    fn arbitrary<'a>(u: &mut arbitrary::Unstructured<'a>) -> arbitrary::Result<Self> {
+        Ok(Self {
+            data: u.arbitrary()?,
+        })
+    }
+}
+
+impl ControlPlane {
+    /// Returns a pseudo-random boolean if the control plane was constructed
+    /// with `arbitrary`.
+    ///
+    /// The default value `false` will always be returned if the
+    /// pseudo-random data is exhausted or the control plane was constructed
+    /// with `default`.
+    pub fn get_decision(&mut self) -> bool {
```

Comment:
One small tweak to the conditional-compilation strategy: I had been thinking that we could have the methods that produce decisions, like `get_decision` here, return a default value (`false` here) as a constant in the non-`chaos`-feature case; then the sites where we use these decisions, like in `MachBuffer`, don't require annotation with conditional compilation. What do you think?

Reply:
I'm thinking about the potential performance impact in release builds, but I guess it's safe to assume the compiler inlines a constant `false` and removes the resulting `if false {}`.

In my view, it would be a nice aspect of the control plane that there is no way to (mis-)use it in regular builds. But I guess that every control plane API needs to have some default output value anyway... and that can probably always be inlined as well? And we can annotate these default-returning functions with `#[inline]`.

What is the downside of conditional compilation at the call sites? It seemed like an easy way to be *really, really sure* nothing bad happens in release builds, but on second thought, it doesn't seem necessary.

Reply:
Indeed, we should be able to trust the branch-folding here.

> What is the downside of conditional compilation at the call sites? 

The main downside is that it spreads the implementation of a conditional decision across distributed points -- the alternative, where everything is wired to a single module where all conditional-compilation logic lies, makes it easier to make changes in the future. (Another example of this principle in action is the `memfd` pooling-allocator mechanism in Wasmtime: when I implemented this in #3697 last year I originally had feature-conditional code in many places, but Alex convinced me to centralize everything into two versions of one module and remove conditionals everywhere else. The result is far cleaner!)

Reply:
I'm assuming this goes for the `Arbitrary` implementation as well, so I removed the conditional compilation [here](https://github.com/bytecodealliance/wasmtime/pull/6039/files#diff-7e77f700326e887cf8ef8327df2692daeb869db9116fb0c44eb80e3f18f1c987R221) too.

The shim control plane's [`Arbitrary` implementation](https://github.com/bytecodealliance/wasmtime/pull/6039/files#diff-39db0fd3f924e4c55e0925a710768bec6cca5922b2ea4d0a842dd2d057e670a9R15-R19) now returns the default without consuming any bytes.